daml/ContingentClaims/Lifecycle.daml

```diff
@@ -17,71 +21,152 @@ import ContingentClaims.Observation (Observation)
 import ContingentClaims.Observable (Interpret(..))
 import ContingentClaims.Util.Recursion (apoCataM)
 import ContingentClaims.Util (pruneZeros')
+import Daml.Control.Recursion (project, embed, distApo)
 import Daml.Control.Monad.Writer (WriterT, runWriterT)
-import Daml.Control.Monad.State (StateT(..), evalStateT)
 import Daml.Control.Monad.MonadWriter (MonadWriter(..))
-import Daml.Control.Monad.MonadState (MonadState(..), modify)
 import Daml.Control.Monad.Trans
 import Daml.Control.Arrow (Kleisli(..))
-import Prelude hiding (sequence, mapA, compare)
+import DA.Traversable (sequence)
+import DA.Foldable (sequence_)
+import Prelude hiding (sequence, mapA, exercise, compare)
 
 type C a = Claim Observation Date Decimal a
 type F a = ClaimF Observation Date Decimal a
 
 deriving instance Eq a => Eq (C a)
 
--- | Returned from a lifecycle operation.
+-- | Returned from a `lifecycle` operation.
 data Result a = Result with
-  remaining : C a
-    -- ^ the tree after lifecycled branches have been pruned.
   pending : [(Decimal, a)]
     -- ^ quantity/asset pairs requiring settlement.
+  remaining : C a
+    -- ^ the tree after lifecycled branches have been pruned.
 
--- | Lifecycle a Claim. Parameters:
--- 1. Function taking key and date, used to lookup an observable on the ledger. `m` will typically be `Update`. It is parametrised so it can be run in a `Script`.
--- 2. Function used when branching in an `Or`. Must return one of the arguments.
--- 3. A set of `Claim`s.
--- 4. A function which, given a date, lifecycles the third argument.
+-- | Collect claims falling due into a list, and return the tree with those nodes pruned.
+-- `m` will typically be `Update`. It is parametrised so it can be run in a `Script`.
+-- The first argument is used to lookup the value of any `Observables`.
+-- Results in a function taking today's date, and returning the pruned tree + pending settlements.
 lifecycle : (Eq a, CanAbort m)
   => (Text -> Date -> m Decimal)
-  -> (C a -> C a -> C a)
   -> C a
   -> Date -> m (Result a)
-lifecycle spot choose
+lifecycle spot
   = runKleisli
-  . fmap (uncurry Result)
-  . flip evalStateT 1.0
+  . fmap (uncurry $ flip Result)
   . runWriterT
-  . apoCataM (pruneZeros') (lifecycle' spot choose)
-
-type Seed m a = WriterT [(Decimal, a)] (StateT Decimal (Kleisli m Date)) (F a (Either (C a) (C a)))
+  . apoCataM pruneZeros' acquireThenSettle
+  . (1.0, )
+  where
+    acquireThenSettle =
+      fmap (fmap (fmap join))
+      . fmap (fmap distE)
+      . fmap join
+      . fmap (fmap sequence)
+      . fmap (fmap (fmap (lifecycle' spot)))
+      . fmap (fmap sequence)
+      . fmap sequence
+      . fmap
+      . fmap (fmap (fmap embed))
+      . fmap (fmap (tsidE : F a (Either (C a) (C a)) -> Either (C a) (F a (C a)) ))
+      . fmap lift
+      $ acquire' spot
```

Comment:
You'll have to talk me through this function :D

Reply:
The composition of functions in `Lifecycle` can be simplified, using identities like `mapA f = sequence . map f`, `fmap f . fmap g = fmap (f . g)` etc. That will also improve performance. I'm raising a ticket for this.